File: frontend/src/services/bookService.js
import api from './api';

export const bookService = {
  // Get all books
  getAll: async (params = {}) => {
    const response = await api.get('/books', { params });
    return response;
  },

  // Get book by ID
  getById: async (id) => {
    const response = await api.get(`/books/${id}`);
    return response;
  },

  // Create new book
  create: async (data) => {
    const response = await api.post('/books', data);
    return response;
  },
//
  // Update book
  update: async (id, data) => {
    const response = await api.put(`/books/${id}`, data);
    return response;
  },

  // Delete book
  delete: async (id) => {
    const response = await api.delete(`/books/${id}`);
    return response;
  },

  // Increment download count
  download: async (id) => {
    const response = await api.post(`/books/${id}/download`);
    return response;
  },

  // Increment view count
  view: async (id) => {
    const response = await api.post(`/books/${id}/view`);
    return response;
  },

  // Get statistics (total books, total downloads)
  getStats: async () => {
    const response = await api.get('/books', {
      params: { per_page: 1 }
    });
    const total = response.meta?.total || 0;
    const totalDownloads = (response.data || []).reduce((sum, book) => sum + (book.downloads || 0), 0);
    return { total, totalDownloads };
  },

  // Saved books
  save: async (data) => {
    const response = await api.post('/saved-books', data);
    return response;
  },

  getSaved: async () => {
    const response = await api.get('/saved-books');
    return response;
  },

  deleteSaved: async (id) => {
    const response = await api.delete(`/saved-books/${id}`);
    return response;
  },
};// logic refinement 4

File: api/backend/internal/services/book_service.go
package services

import (
	"errors"

	"e-library/backend/internal/models"
	"e-library/backend/internal/repositories"
)

type BookService struct {
	bookRepo     *repositories.BookRepository
	categoryRepo *repositories.CategoryRepository
}

func NewBookService(
	bookRepo *repositories.BookRepository,
	categoryRepo *repositories.CategoryRepository,
) *BookService {
	return &BookService{
		bookRepo:     bookRepo,
		categoryRepo: categoryRepo,
	}
}

//  GET ALL BOOKS 
func (s *BookService) GetAllBooks(
	filters map[string]interface{},
) ([]models.Book, *models.Meta, error) {

	books, err := s.bookRepo.FindAll(filters)
	if err != nil {
		return nil, nil, err
	}

	total, err := s.bookRepo.Count(filters)
	if err != nil {
		return books, nil, nil
	}
//
	page := 1
	perPage := 12

	if p, ok := filters["page"].(int); ok && p > 0 {
		page = p
	}
	if pp, ok := filters["per_page"].(int); ok && pp > 0 {
		perPage = pp
	}

	totalPages := (total + perPage - 1) / perPage

	meta := &models.Meta{
		Total:      total,
		Page:       page,
		PerPage:    perPage,
		TotalPages: totalPages,
	}

	return books, meta, nil
}

//  GET BOOK BY ID 
func (s *BookService) GetBookByID(id int) (*models.Book, error) {
	book, err := s.bookRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if book == nil {
		return nil, errors.New("book not found")
	}
	return book, nil
}

//  CREATE BOOK 
func (s *BookService) CreateBook(book *models.Book) error {

	if book.CategoryID != nil {
		cat, err := s.categoryRepo.FindByID(*book.CategoryID)
		if err != nil {
			return err
		}
		if cat == nil {
			return errors.New("category not found")
		}
	}

	return s.bookRepo.Create(book)
}

//  UPDATE BOOK 
func (s *BookService) UpdateBook(id int, book *models.Book) error {

	existing, err := s.bookRepo.FindByID(id)
	if err != nil {
		return err
	}
	if existing == nil {
		return errors.New("book not found")
	}

	if book.CategoryID != nil {
		cat, err := s.categoryRepo.FindByID(*book.CategoryID)
		if err != nil {
			return err
		}
		if cat == nil {
			return errors.New("category not found")
		}
	}

	return s.bookRepo.Update(id, book)
}

//  DELETE BOOK 
func (s *BookService) DeleteBook(id int) error {
	return s.bookRepo.Delete(id)
}

//  INCREMENT DOWNLOAD 
func (s *BookService) IncrementDownload(id int) error {
	return s.bookRepo.IncrementDownloads(id)
}

//  INCREMENT VIEW 
func (s *BookService) IncrementView(id int) error {
	return s.bookRepo.IncrementViews(id)
}
